extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java

```diff
@@ -14,33 +16,62 @@ public class TenantConfigBean {
     private final Map<String, TenantConfigContext> staticTenantsConfig;
     private final Map<String, TenantConfigContext> dynamicTenantsConfig;
     private final TenantConfigContext defaultTenant;
-    private final Function<OidcTenantConfig, Uni<TenantConfigContext>> tenantConfigContextFactory;
+    private final TenantContextFactory tenantContextFactory;
+    private final Map<String, TenantConfigContext> unmodifiableDynamicTenants;
+
+    @FunctionalInterface
+    public interface TenantContextFactory {
+        Uni<TenantConfigContext> create(OidcTenantConfig oidcTenantConfig);
+    }
 
     public TenantConfigBean(
             Map<String, TenantConfigContext> staticTenantsConfig,
-            Map<String, TenantConfigContext> dynamicTenantsConfig,
             TenantConfigContext defaultTenant,
-            Function<OidcTenantConfig, Uni<TenantConfigContext>> tenantConfigContextFactory) {
+            TenantContextFactory tenantContextFactory) {
         this.staticTenantsConfig = Map.copyOf(staticTenantsConfig);
-        this.dynamicTenantsConfig = dynamicTenantsConfig;
+        this.dynamicTenantsConfig = new ConcurrentHashMap<>();
+        this.unmodifiableDynamicTenants = Collections.unmodifiableMap(dynamicTenantsConfig);
         this.defaultTenant = defaultTenant;
-        this.tenantConfigContextFactory = tenantConfigContextFactory;
+        this.tenantContextFactory = tenantContextFactory;
+    }
+
+    public Uni<TenantConfigContext> createDynamicTenantContext(OidcTenantConfig oidcConfig) {
+        var tenantId = oidcConfig.tenantId.orElseThrow();
+
+        var tenant = dynamicTenantsConfig.get(tenantId);
+        if (tenant != null) {
+            return Uni.createFrom().item(tenant);
+        }
+
+        return tenantContextFactory.create(oidcConfig).onItem().transform(
+                new Function<TenantConfigContext, TenantConfigContext>() {
+                    @Override
+                    public TenantConfigContext apply(TenantConfigContext t) {
+                        dynamicTenantsConfig.putIfAbsent(tenantId, t);
+                        return t;
+                    }
+                });
     }
 
     public Map<String, TenantConfigContext> getStaticTenantsConfig() {
         return staticTenantsConfig;
     }
 
+    public TenantConfigContext getStaticTenant(String tenantId) {
+        return staticTenantsConfig.get(tenantId);
+    }
+
     public TenantConfigContext getDefaultTenant() {
         return defaultTenant;
     }
 
-    public Function<OidcTenantConfig, Uni<TenantConfigContext>> getTenantConfigContextFactory() {
-        return tenantConfigContextFactory;
+    @SuppressWarnings("unused") // retained for ABI compatibility
```

Comment:
@snazy I think we can remove it, this is not something users are expected to deal with directly, it is in the runtime package as well, and we don't particularly recommend using `TenantConfigBean` directly

Reply:
done